### PR TITLE
use `Plutarch.Prelude` Mixin

### PR DESCRIPTION
**Pull request by SeungheonOh**

### CODE REVIEW

**Review by chfanghr — APPROVED** [review]
🚀

**Comment by peter-mlabs** [comment]
@kozross, how do you feel about allowing `import Plutarch.Prelude` without an explicit import list, or the way we do for dApps with `mixins` and an implicit prelude?

**Comment by kozross** [comment]
I feel that `Plutatch.Prelude` is not a full replacement for the `base` one, and therefore should be treated like any other import.

If we want a replacement `Prelude`, which may or may not include identifiers coming from Plutarch, that's a separate story IMHO, since there's definitely some disagreement over what it should, and shouldn't, include. I certainly have some thoughts of my own, but they'd need a full writeup.

**Comment by peter-mlabs** [comment]
FTR, in all the non Liqwid Lib's repos (as well as [plutus-scaffold](https://github.com/mlabs-haskell/plutus-scaffold/blob/master/onchain/template-project.cabal), we use

```
  mixins:
    base hiding (Prelude),
    pprelude (PPrelude as Prelude)
``` 

in the  cabal and

```hs
module PPrelude (
  module Prelude,
  module Plutarch.Prelude,
) where

import Plutarch.Prelude
import Prelude
```

as `PPrelude`. I think this would be good to adopt if there's no strong opposition.

**Comment by kozross** [comment]
@peter-mlabs I'm not opposed - there are no name clashes or major surprises, and the Plutarch prelude is its intended (primary) API.

**Comment by SeungheonOh** [comment]
Okay, It seems like we are going to use `PPrelude` I'll update this PR.

File: src/Plutarch/Extra/FixedDecimal.hs [hunk 1]
```diff
@@ -23,9 +23,7 @@ import GHC.TypeLits (KnownNat, Nat, natVal)
 import qualified Generics.SOP as SOP
 import Plutarch.Api.V1 (AmountGuarantees, KeyGuarantees, PValue)
 import Plutarch.Api.V1.Value (psingletonValue)
-import Plutarch.Bool (PEq, POrd, PPartialOrd)
 import Plutarch.Extra.Function (pflip)
-import Plutarch.Integer (PInteger, PIntegral (pdiv))
 import Plutarch.Num (PNum (..))
```
Comment (kozross) [inline src/Plutarch/Extra/FixedDecimal.hs:27]: A bit stale, but proper imports.

File: src/Plutarch/Extra/FixedDecimal.hs [hunk 2]
```diff
@@ -23,9 +23,7 @@ import GHC.TypeLits (KnownNat, Nat, natVal)
...
 import qualified Plutarch.Numeric.Additive as A (
     AdditiveMonoid (..),
```
Comment (kozross) [inline src/Plutarch/Extra/FixedDecimal.hs:29]: Also here.

File: src/Plutarch/Extra/FixedDecimal.hs [hunk 3]
```diff
@@ -35,7 +33,12 @@ import Plutarch.Prelude (
     DerivePlutusType (..),
```
Comment (kozross) [inline src/Plutarch/Extra/FixedDecimal.hs:33]: And here.

File: src/Plutarch/Extra/List.hs [hunk 4]
```diff
@@ -24,33 +24,49 @@ module Plutarch.Extra.List (
...
+import Plutarch.Extra.TermCont (pletC)
+import Plutarch.Prelude (
+    PAsData,
+    PBool (..),
+    PBuiltinPair,
+    PEq (..),
+    PInteger,
+    PIsData,
+    PIsListLike,
+    PListLike (..),
```
Comment (kozross) [inline src/Plutarch/Extra/List.hs:41]: Proper imports please.

File: src/Plutarch/Extra/Map/Sorted.hs [hunk 5]
```diff
@@ -4,35 +4,38 @@
...
+import Plutarch.Extra.Map (pkeys)
+import Plutarch.Extra.TermCont (pletC)
+import Plutarch.Prelude (
+    PAsData,
+    PBool,
+    PBuiltinList,
+    PBuiltinPair,
+    PEq ((#==)),
+    PIsData,
+    PListLike (..),
```
Comment (kozross) [inline src/Plutarch/Extra/Map/Sorted.hs:18]: Proper imports.

File: src/Plutarch/Extra/These.hs [hunk 6]
```diff
@@ -42,12 +33,25 @@ import Plutarch.Extra.Traversable (PTraversable (ptraverse, ptraverse_))
...
     DPTStrat,
     DerivePlutusType,
+    PAsData,
+    PBool (PFalse, PTrue),
+    PDataRecord,
+    PEq,
+    PIsData,
+    PLabeledType ((:=)),
+    POrd,
+    PPartialOrd (..),
```
Comment (kozross) [inline src/Plutarch/Extra/These.hs:43]: Proper imports.

**Review by kozross — CHANGES_REQUESTED** [review]

**Comment by SeungheonOh** [comment]
Okay using mix-ins looks so much cleaner.

**Comment by peter-mlabs** [comment]
Added `-Wmissing-import-lists` to the cabal and `{-# OPTIONS_GHC -Wno-missing-import-lists #-}` to the files with `import Prelude hiding (..,)` statements.

**Review by peter-mlabs — APPROVED** [review]
All should be good now, thanks!

File: src/Plutarch/Extra/AssetClass.hs [hunk 7]
```diff
@@ -3,6 +3,7 @@
 {-# LANGUAGE TypeApplications #-}
 {-# LANGUAGE TypeFamilies #-}
 {-# LANGUAGE UndecidableInstances #-}
+{-# OPTIONS_GHC -Wno-missing-import-lists #-}
```
Comment (kozross) [inline src/Plutarch/Extra/AssetClass.hs:6]: A comment explaining why this is off would be good I think, for future reference if nothing else.

File: src/Plutarch/Extra/Bind.hs [hunk 8]
```diff
@@ -1,6 +1,7 @@
 {-# LANGUAGE QuantifiedConstraints #-}
 {-# LANGUAGE TypeApplications #-}
 {-# LANGUAGE UndecidableInstances #-}
+{-# OPTIONS_GHC -Wno-missing-import-lists #-}
```
Comment (kozross) [inline src/Plutarch/Extra/Bind.hs:4]: Same here.

File: src/Plutarch/Extra/Map.hs [hunk 9]
```diff
@@ -1,6 +1,7 @@
 {-# LANGUAGE PartialTypeSignatures #-}
 {-# LANGUAGE QuantifiedConstraints #-}
 {-# LANGUAGE ViewPatterns #-}
+{-# OPTIONS_GHC -Wno-missing-import-lists #-}
```
Comment (kozross) [inline src/Plutarch/Extra/Map.hs:4]: Also here.

File: src/Plutarch/Extra/Traversable.hs [hunk 10]
```diff
@@ -1,5 +1,6 @@
 {-# LANGUAGE QuantifiedConstraints #-}
 {-# LANGUAGE TypeApplications #-}
+{-# OPTIONS_GHC -Wno-missing-import-lists #-}
```
Comment (kozross) [inline src/Plutarch/Extra/Traversable.hs:3]: And here.

File: test/Spec.hs [hunk 11]
```diff
@@ -5,7 +5,8 @@ import Test.Tasty (defaultMain, testGroup)
 
 import qualified Spec.Extra.List as List
 import qualified Spec.Extra.Map as Map
-import qualified Spec.Extra.MultiSig as MultiSig
+
+-- import qualified Spec.Extra.MultiSig as MultiSig
```
Comment (kozross) [inline test/Spec.hs:9]: Should this be removed?

File: test/Spec/Extra/Map/Sorted.hs [hunk 12]
```diff
@@ -26,11 +26,7 @@ import qualified GHC.Generics as GHC
 
 --------------------------------------------------------------------------------
 
-import Plutarch (Term, phoistAcyclic, plam, (#), type (:-->))
-import Plutarch.Api.V2.AssocMap (KeyGuarantees (..), PMap)
-import Plutarch.Bool (PBool)
-import Plutarch.Builtin (PBuiltinPair, pfstBuiltin, psndBuiltin)
-import Plutarch.Integer (PInteger)
+import Plutarch.Api.V1.AssocMap (KeyGuarantees (..), PMap)
```
Comment (kozross) [inline test/Spec/Extra/Map/Sorted.hs:29]: Wildcard imports shouldn't be used.

**Review by kozross — CHANGES_REQUESTED** [review]

File: src/Plutarch/Extra/FixedDecimal.hs [hunk 13]
```diff
@@ -18,39 +18,16 @@ module Plutarch.Extra.FixedDecimal (
...
-import Plutarch.Api.V2 (AmountGuarantees, KeyGuarantees)
-import Plutarch.Bool (PEq, POrd, PPartialOrd)
+import Plutarch.Api.V1 (AmountGuarantees, KeyGuarantees, PValue)
 import Plutarch.Extra.Function (pflip)
 import Plutarch.Extra.Value (psingletonValue)
-import Plutarch.Integer (PInteger, PIntegral (pdiv))
-import Plutarch.Num (PNum (..))
+import Plutarch.Num (PNum (pfromInteger, (#*)))
 import qualified Plutarch.Numeric.Additive as A (
     AdditiveMonoid (..),
```
Comment (kozross) [inline src/Plutarch/Extra/FixedDecimal.hs:28]: Imports.

File: src/Plutarch/Extra/FixedDecimal.hs [hunk 14]
```diff
@@ -18,39 +18,16 @@ module Plutarch.Extra.FixedDecimal (
...
     AdditiveSemigroup (..),
```
Comment (kozross) [inline src/Plutarch/Extra/FixedDecimal.hs:29]: Also here.

File: src/Plutarch/Extra/Identity.hs [hunk 15]
```diff
@@ -7,25 +7,8 @@ module Plutarch.Extra.Identity (
     PIdentity (..),
```
Comment (kozross) [inline src/Plutarch/Extra/Identity.hs:7]: Imports.

**Review by kozross — CHANGES_REQUESTED** [review]

**Review by kozross — APPROVED** [review]

**Pull request merged**